### PR TITLE
8340721: Clarify special case handling of unboxedType and getWildcardType

### PR DESCRIPTION
**Pull request by jddarcy**
First piece of a larger set of updates that may be made for JDK-8055219 to clarify the exceptional behavior of `javax.lang.model.util.Types`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8340979](https://bugs.openjdk.org/browse/JDK-8340979) to be approved

### Issues
 * [JDK-8340721](https://bugs.openjdk.org/browse/JDK-8340721): Clarify special case handling of unboxedType and getWildcardType (**Sub-task** - P4)
 * [JDK-8340979](https://bugs.openjdk.org/browse/JDK-8340979): Clarify special case handling of unboxedType and getWildcardType (**CSR**)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21193/head:pull/21193` \
`$ git checkout pull/21193`

Update a local copy of the PR: \
`$ git checkout pull/21193` \
`$ git pull https://git.openjdk.org/jdk.git pull/21193/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21193`

View PR using the GUI difftool: \
`$ git pr show -t 21193`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21193.diff">https://git.openjdk.org/jdk/pull/21193.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21193#issuecomment-2375391880)

### CODE REVIEW

**Comment by jddarcy** [comment]
> First piece of a larger set of updates that may be made for JDK-8055219 to clarify the exceptional behavior of `javax.lang.model.util.Types`.

Please also review the CSR [JDK-8340979](https://bugs.openjdk.org/browse/JDK-8340979).

I may augment the regression test to test unions types. Unions types come about from multi-catch statements so getting a hold of one involves a trip with the tree API and not just the standard annotation processing APIs.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        // non-reference: ExecutableType, IntersectionType, NoType, PrimitiveType, UnionType, WildcardType
+        var invalidInputs = List.of(objectType, stringType, arrayType,
+                                    executableType, intersectionType,
+                                    noTypeVoid, noTypeNone, noTypePackage, noTypeModule, nullType,
+                                    primitiveType, /*unionType, */ wildcardType);
+
+        for (TypeMirror tm : invalidInputs) {
+            try {
+                PrimitiveType pt = types.unboxedType(tm);
+            } catch(IllegalArgumentException iae) {
```
Comment (pavelrappo) [inline test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:149]: Here and elsewhere in this file: if some of the calls didn't throw IAE, the test would still pass. Is this your goal?

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        // Reference types are ArrayType, DeclaredType, ErrorType, NullType, TypeVariable
+        // non-reference: ExecutableType, IntersectionType, NoType, PrimitiveType, UnionType, WildcardType
+        var invalidInputs = List.of( arrayType,
+                                    executableType, intersectionType,
+                                    noTypeVoid, noTypeNone, noTypePackage, noTypeModule, nullType,
+                                    primitiveType, /*unionType, */ wildcardType);
+
+        for (TypeMirror tm : invalidInputs) {
+            try {
+                WildcardType wc1 = types.getWildcardType(tm,   null);
```
Comment (pavelrappo) [inline test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:171]: Calling getWildcardType for an array type does not throw IAE; is this expected?

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java [hunk 3]
```diff
@@ -50,6 +50,11 @@
  * <p><b>Compatibility Note:</b> Methods may be added to this interface
  * in future releases of the platform.
  *
+ * @apiNote
+ * In the reference implementation, handling {@linkplain ErrorType
+ * error types} generally does not cause an {@code
+ * UnsupportedOperationException} from the methods in this interface.
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/Types.java:56]: Yes, meant IAE; will correct in the next push.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java (hunk 2 above)
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:171]: Will correct in next push; thanks.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java (hunk 1 above)
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:148]: Good catch; corrected.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 4]
```diff
@@ -161,20 +160,21 @@ void testUnboxedType() {
...
             try {
                 WildcardType wc1 = types.getWildcardType(tm,   null);
+                throw new RuntimeException("Should not reach " + tm);
             } catch(IllegalArgumentException iae) {
                 ; // Expected
             }
 
             try {
                 WildcardType wc2 = types.getWildcardType(null, tm);
+                throw new RuntimeException("Should not reach" + tm);
```
Comment (pavelrappo) [inline test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:177]:
```suggestion
                throw new RuntimeException("Should not reach " + tm);
```

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 5]
```diff
@@ -146,6 +144,7 @@ void testUnboxedType() {
         for (TypeMirror tm : invalidInputs) {
             try {
                 PrimitiveType pt = types.unboxedType(tm);
+                throw new RuntimeException("Should not reach" + tm);
```
Comment (pavelrappo) [inline test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:147]:
```suggestion
                throw new RuntimeException("Should not reach " + tm);
```

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 6]
```diff
@@ -161,20 +160,21 @@ void testUnboxedType() {
     void testGetWildcardType() {
         // Reference types are ArrayType, DeclaredType, ErrorType, NullType, TypeVariable
         // non-reference: ExecutableType, IntersectionType, NoType, PrimitiveType, UnionType, WildcardType
-        var invalidInputs = List.of( arrayType,
```
Comment (pavelrappo) [inline test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java]: Do we need to change `getWildcardType`'s behaviour for `ArrayType` or specification? `ArrayType` is a reference type, and `getWildcardType` is now specified to throw IAE for all reference types.

File: src/java.compiler/share/classes/javax/lang/model/util/Types.java [hunk 7]
```diff
@@ -268,7 +275,10 @@ public interface Types {
      *
      * @param extendsBound  the extends (upper) bound, or {@code null} if none
      * @param superBound    the super (lower) bound, or {@code null} if none
-     * @throws IllegalArgumentException if bounds are not valid
+     *
+     * @throws IllegalArgumentException if bounds are not valid,
+     * including for types that are <em>not</em> {@linkplain ReferenceType
```
Comment (mcimadamore) [inline src/java.compiler/share/classes/javax/lang/model/util/Types.java:280]: I find "including for types" hard to parse and connect to the previous clause. Did you mean "e.g. if if the bounds are not reference types" ?

Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/Types.java:280]: Okay; I'll rephrase this to be more clear. Something like:

"throws IllegalArgumentException - if bounds are not valid. Invalid bounds include all types that are not reference types."

The intention of the phrasing is meant to imply there are reference type arguments that could be invalid, per any restrictions from the JLS. Thanks.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java (hunk 6 above)
Comment (pavelrappo) [reply test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java]: Oops, disregard that previous comment. My bad, was sloppily reviewing this part.

**Review by pavelrappo — APPROVED** [review]
Now looks good.

File: test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+ */
+public class TestInvalidInputs extends JavacTestingAbstractProcessor {
+
+    // Reference types are ArrayType, DeclaredType, ErrorType, NullType, and TypeVariable
+
+    private TypeMirror       objectType;
+    private TypeMirror       stringType;
+    private ArrayType        arrayType;
+    private DeclaredType     declaredType;
+    // private ErrorType        errorType; // skip for now
```
[the remark this answers is not in the document]
Comment (jddarcy) [reply test/langtools/tools/javac/processing/model/util/types/TestInvalidInputs.java:50]: I think going that reliably could take a bit of extra effort and need to step outside of the one-file-for-a-test methodology. For example, have an "Erroneous.java" test that was only compiled during the annotation processioning phase and not when the annotation processor itself is built.

**Review by mcimadamore — APPROVED** [review]
